topi/python/topi/arm_cpu/conv2d.py

```diff
@@ -499,6 +508,70 @@ def _schedule_winograd(cfg, s, output, last):
         s[output].compute_inline()
 
 
+@autotvm.register_topi_compute(conv2d, 'arm_cpu', ['winograd_nnpack'])
+def conv2d_arm_cpu_winograd_nnpack(
+        cfg, data, kernel, strides, padding, dilation, layout, out_dtype):
+    """ TOPI compute callback. Use winograd NNPACK template """
+    N, CI, IH, IW = get_const_tuple(data.shape)
+
+    if isinstance(dilation, int):
+        dilation_h = dilation_w = dilation
+    else:
+        dilation_h, dilation_w = dilation
+    assert (dilation_h, dilation_w) == (1, 1)
+    assert len(kernel.shape) == 4
+    CO, _, KH, KW = get_const_tuple(kernel.shape)
+    HSTR, WSTR = strides if isinstance(strides, (tuple, list)) else (strides, strides)
+    HPAD, WPAD, _, _ = get_pad_tuple(padding, kernel)
+
+    assert layout == 'NCHW'
+    assert KH == 3 and KW == 3 and HPAD == 1 and WPAD == 1 and HSTR == 1 and WSTR == 1
+    H = (IH + 2 * HPAD - 3) // HSTR + 1
+    W = (IW + 2 * WPAD - 3) // WSTR + 1
+
+    cfg.define_knob(
+        'winograd_nnpack_algorithm',
+        [
+            tvm.contrib.nnpack.ConvolutionAlgorithm.WT_8x8,
+            tvm.contrib.nnpack.ConvolutionAlgorithm.WT_8x8_FP16
+        ]
+    )
+
+    assert N == 1
+    with tvm.tag_scope("winograd_nnpack_conv2d_weight_transform"):
+        transformed_kernel = tvm.contrib.nnpack.convolution_inference_weight_transform(
+            kernel, algorithm=cfg['winograd_nnpack_algorithm'].val)
+        if autotvm.GLOBAL_SCOPE.in_tuning:
+            transformed_kernel = tvm.compute(transformed_kernel.shape, lambda *args: 0.0)
+
+    with tvm.tag_scope("winograd_nnpack_conv2d_output"):
+        output = tvm.contrib.nnpack.convolution_inference_without_weight_transform(
+            data, transformed_kernel,
+            bias=None,
+            padding=[HPAD, HPAD, WPAD, WPAD],
+            stride=[HSTR, WSTR],
+            algorithm=cfg['winograd_nnpack_algorithm'].val)
+
+    # we have to manually assign effective GFLOP for winograd
+    cfg.add_flop(2 * N * CI * H * W * KH * KW * CO)
```

Comment:
Out of curiosity, is it because nnpack needs this info to decide if winograd should kick in?

Reply:
There are two reasons:
1) NNPACK is an extern op, `compute_flops` does not know how to compute the flops for extern ops
2) The winograd implementation reduces the algorithmic complexity by reducing the number of multiplication operators, so here we care about the "fake GFlops" (the number of multiplys as if we did a direct conv) so later AutoTVM can choose the best out of direct, winograd, winograd_nnpack.